cpp/daal/src/algorithms/service_kernel_math.h

```diff
@@ -263,6 +270,389 @@ class EuclideanDistances : public PairwiseDistances<FPType, cpu>
+
+template <>
+float MinkowskiDistances<float, avx512>::computeDistance(const float * x, const float * y, const size_t n)
+{
+    daal::internal::mkl::MklMath<float, avx512> math;
+    float * tmp = new float[16];
+    float d     = 0.0;
+
+    __m512 * ptr512x = (__m512 *)x;
+    __m512 * ptr512y = (__m512 *)y;
+
+    if (_p == 1.0)
+    {
+        for (size_t i = 0; i < n / 16; ++i)
+        {
+            d += _mm512_reduce_add_ps(_mm512_abs_ps(_mm512_sub_ps(ptr512x[i], ptr512y[i])));
+        }
+
+        for (size_t i = (n / 16) * 16; i < n; ++i)
+        {
+            d += math.sFabs(x[i] - y[i]);
+        }
+
+        return d;
+    }
+    else
+    {
+        for (size_t i = 0; i < n / 16; ++i)
+        {
+            _mm512_storeu_ps(tmp, _mm512_abs_ps(_mm512_sub_ps(ptr512x[i], ptr512y[i])));
+            math.vPowx(4, tmp, _p, tmp);
+            d += _mm512_reduce_add_ps(_mm512_loadu_ps(tmp));
+        }
+
+        delete[] tmp;
+
+        for (size_t i = (n / 16) * 16; i < n; ++i)
+        {
+            d += math.sPowx(math.sFabs(x[i] - y[i]), _p);
+        }
+
+        if (!_powered) return math.sPowx(d, 1.0 / _p);
+
+        return d;
+    }
+}
+
+template <>
+double MinkowskiDistances<double, avx512>::computeDistance(const double * x, const double * y, const size_t n)
+{
+    daal::internal::mkl::MklMath<double, avx512> math;
+
+    double d          = 0.0;
+    double * tmp      = new double[8];
+    __m512d * ptr512x = (__m512d *)x;
+    __m512d * ptr512y = (__m512d *)y;
+
+    if (_p == 1.0)
+    {
+        for (size_t i = 0; i < n / 8; ++i)
+        {
+            d += _mm512_reduce_add_pd(_mm512_abs_pd(_mm512_sub_pd(ptr512x[i], ptr512y[i])));
+        }
+
+        for (size_t i = (n / 8) * 8; i < n; ++i)
+        {
+            d += math.sFabs(x[i] - y[i]);
+        }
+
+        return d;
+    }
+    else
+    {
+        for (size_t i = 0; i < n / 8; ++i)
+        {
+            _mm512_storeu_pd(tmp, _mm512_abs_pd(_mm512_sub_pd(ptr512x[i], ptr512y[i])));
+            math.vPowx(8, tmp, _p, tmp);
+            d += _mm512_reduce_add_pd(_mm512_loadu_pd(tmp));
+        }
+
+        delete[] tmp;
+
+        for (size_t i = (n / 8) * 8; i < n; ++i)
+        {
+            d += math.sPowx(math.sFabs(x[i] - y[i]), _p);
+        }
+
+        if (!_powered) return math.sPowx(d, 1.0 / _p);
+
+        return d;
+    }
+}
+
+template <>
+float ChebychevDistances<float, avx512>::computeDistance(const float * x, const float * y, const size_t n)
+{
+    daal::internal::mkl::MklMath<float, avx512> math;
+    float * tmp = new float[16];
+    float d     = 0.0;
+
+    __m512 * ptr512x = (__m512 *)x;
+    __m512 * ptr512y = (__m512 *)y;
+
+    __m512 tmp512 = _mm512_abs_ps(_mm512_sub_ps(ptr512x[0], ptr512y[0]));
+
+    for (size_t i = 1; i < n / 16; ++i)
+    {
+        tmp512 = _mm512_max_ps(tmp512, _mm512_abs_ps(_mm512_sub_ps(ptr512x[i], ptr512y[i])));
+    }
+
+    _mm512_storeu_ps(tmp, tmp512);
+
+    for (size_t i = 0; i < 16; ++i)
```

Comment:
Please use variable called `vec_size` or smth like that instead of plain magic numbers like 16

Reply:
applied